hermes/probe/checknil/check_nil.go

```diff
@@ -0,0 +1,332 @@
+
+	var readErr error
+	file.contents, readErr = ioutil.ReadAll(reader)
+	if readErr != nil {
+		return nil, probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("readNilFile(): unable to read nil file: %w", readErr))
+	}
+
+	logger.Infof("readNilFile(): nil file read complete.")
+
+	return file, probe.NewProbeError(metrics.Success, nil)
+}
+
+// verifyNilFile verifies that the contents of the nil file match its checksum
+// and that the journal matches the existing in-memory proto.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- nilFile: a nilFile struct with the file name and contents recorded
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func verifyNilFile(ctx context.Context, file *nilFile, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	hash := sha1.New()
+	if _, err := io.Copy(hash, file.reader()); err != nil {
+		return probe.NewProbeError(metrics.ProbeFailed, fmt.Errorf("verifyNilFile(): unable to compute checksum of file contents: %w", err))
+	}
+	file.hash = fmt.Sprintf("%x", hash.Sum(nil))
+
+	if !file.validateChecksum() {
+		return probe.NewProbeError(metrics.FileCorrupted, fmt.Errorf("verifyNilFile(): checksum of nil file did not match filename"))
+	}
+
+	journal := &sjpb.StateJournal{}
+	if err := proto.Unmarshal(file.contents, journal); err != nil {
+		return probe.NewProbeError(metrics.FileReadFailure, fmt.Errorf("verifyNilFile(): unable to unmarshal nil file proto: %w", err))
+	}
+
+	logger.Infof("verifyNilFile(): unmarshal complete for nil file proto.")
+
+	if err := resolveIntent(ctx, journal, target, client, logger); err != nil {
+		return err
+	}
+
+	if diff := cmp.Diff(target.Journal, journal, protocmp.Transform()); diff != "" {
+		return probe.NewProbeError(metrics.FileCorrupted, fmt.Errorf("verifyNilFile(): journal contents mismatch (-want +got):\n%s", diff))
+	}
+
+	logger.Infof("verifyNilFile(): nil file contents verified against in-memory struct.")
+
+	return probe.NewProbeError(metrics.Success, nil)
+}
+
+// resolveIntent verifies and completes the intent stored in the state journal passed.
+// Arguments:
+//	- ctx: context so this probe can be cancelled if needed.
+//	- journal: the StateJournal read from the storage system.
+//	- target: target run information.
+//	- client: initialised storage client for this target system.
+//	- logger: logger associated with the probe calling this function.
+// Returns:
+//	- err: error exit statuses are well defined in metrics.go
+func resolveIntent(ctx context.Context, journal *sjpb.StateJournal, target *probe.Target, client stiface.Client, logger *logger.Logger) *probe.ProbeError {
+	if journal.Intent != nil {
+		fileID, err := strconv.Atoi(journal.Intent.Filename[len("Hermes_"):len("Hermes_ID")])
```

Comment:
The Nil file (now Journal) will never be stored as part of the StateJournal struct, as it would be self-referential and we could not checksum it correctly. Therefore, this is not meant to support the nil file as input. 

I will add a check for out of range IDs, but this should't occur as both the create and delete operations have strict bounds on the IDs that they can operate on.